### ISSUE TITLE
Plugin fails on Android project when Gradle Instant Execution is enabled

### ISSUE THREAD

**Issue by eskatos** [google/protobuf-gradle-plugin#396]
Run a build on an android project that uses the protobug plugin with `-Dorg.gradle.unsafe.instant-execution=true`, it fails with the following report:

```
▼ task:subproject:generateDebugProto:clipboard: of type com.google.protobuf.gradle.GenerateProtoTask:clipboard:
 ▼ fieldactions:clipboard: of com.google.protobuf.gradle.GenerateProtoTask:clipboard:
  ▼ bean of type org.gradle.api.internal.AbstractTask$ClosureTaskAction:clipboard:
   ▼ fieldclosure:clipboard: of org.gradle.api.internal.AbstractTask$ClosureTaskAction:clipboard:
    ▼ bean of type com.google.protobuf.gradle.ToolsLocator$_registerDependencyWithTasks_closure2:clipboard:
     ▼ fielddep:clipboard: of com.google.protobuf.gradle.ToolsLocator$_registerDependencyWithTasks_closure2:clipboard:
      ▼ bean of type groovy.lang.Reference:clipboard:
       ▼ fieldvalue:clipboard: of groovy.lang.Reference:clipboard:
        ▼ bean of type org.gradle.api.internal.artifacts.dependencies.DefaultExternalModuleDependency:clipboard:
         ▼ fieldattributesFactory:clipboard: of org.gradle.api.internal.artifacts.dependencies.DefaultExternalModuleDependency:clipboard:
          ▼ bean of type org.gradle.api.internal.attributes.DefaultImmutableAttributesFactory:clipboard:
           ▼ fieldisolatableFactory:clipboard: of org.gradle.api.internal.attributes.DefaultImmutableAttributesFactory:clipboard:
            ▼ bean of type org.gradle.internal.snapshot.impl.DefaultValueSnapshotter
```

https://github.com/google/protobuf-gradle-plugin/pull/381 fixed the plugin for Java/Kotlin only builds but there are more issues with Android builds.

**Comment by ejona86** [comment]
That doesn't actually include any error. What is it saying is wrong?

**Comment by eskatos** [comment]
Sorry my report wasn't clear. I'll try to clarify.

I just tried to run `ProtobufAndroidPluginTest` using latest Gradle nightly (`6.5-20200415221304+0000`) and latest AGP alpha (`4.1.0-alpha05`) with instant execution enabled and here is the report I get:

![image](https://user-images.githubusercontent.com/132773/79446207-277a1180-7fde-11ea-9223-dda9c2dc587f.png)

You can download and look at the report [instant-execution-report.zip](https://github.com/google/protobuf-gradle-plugin/files/4486469/instant-execution.zip)

The problems under _Gradle runtime_ are from AGP itself and should be fixed soon.

Instant execution serializes and reuses the Gradle task graph. The two problems reported for the `GenerateProtoTask` points at state held by the task instances that are not supported by instant execution: `SourceSet` instances, and AGP `Variant` instances.

In order to fix the protobuf plugin so it supports instant execution on Android builds it needs to be reworked so that `GenerateProtoTask` doesn't hold references to those instances.

**Comment by voidzcy** [comment]
The original issue should have been fixed by #408. The remaining problems with configuration cache (instant execution) are tracked by #409 and #419. So, closing this one.
